lib/utils.bash

```diff
@@ -7,6 +7,45 @@ GREP_COLORS=
 
 ASDF_DIR=${ASDF_DIR:-''}
 ASDF_DATA_DIR=${ASDF_DATA_DIR:-''}
+BASH_MAJOR_VERSION=${BASH_VERSION%%.*}
+
+fast_basename() {
```

Comment:
This can be changed to:?

```sh
local path="$1"
path=${path%/}
path=${path##*/}
printf '%s\n' "$REPLY"
```

The extra tests seem redundant as Bash should perform them internally.

I think it's _very_ worth pursuing having a `REPLY`-like pattern as shown [here](https://github.com/asdf-vm/asdf/blob/d81b81f9de2dc5961624464df04cef7cafae588c/lib/utils.bash#L852) to completely eliminate the subshell (I strongly prefer the variable being called `REPLY`, though). In my experience this gives gigantic perf wins, something similar as what it seems you have alluded to in your other comment. I'm not sure how on-board the core-team is with this, though.

Reply:
Agreed that the additional subshells can be removed, ty for this recommendation.

Re: `$REPLY`, I thought that was specifically for [read](https://tldp.org/LDP/abs/html/internalvariables.html)? Using your code sample as-is resulted in an empty output for me.

Reply:
Yeah it is for `read` and also `select`, I'm borrowing the convention (more info [here](https://github.com/hyperupcall/awesome-basalt/blob/main/BEST_PRACTICES.md#functions)). It will work if the function looks something like this:

```sh
fast_basename() {
  unset -v REPLY; REPLY=

  local path="$1"
  path=${path%/}
  path=${path##*/}
  REPLY=$path
}
```
And is called like this:

```sh
fast_basename "$executable_path"
local cmd="$REPLY"
```

The code sample I provided was for existing subshell code, not the `REPLY`-thing.

Reply:
For perf test, you have to include the capture of the variable as a part of the test. These were my results:


<details>

<summary>script.sh</summary>

```sh
#!/usr/bin/env bash

fast_basename_work() {
	local path="$1"
	path=${path%/}
	printf '%s\n' "{path##*/}"
}

slow_basename() {
	unset -v REPLY
	REPLY=$(basename "$1")
}

fast_basename() {
	unset -v REPLY
	REPLY=$(fast_basename_work "$1")
}

faster_basename() {
	unset -v REPLY; REPLY=
	local path="$1"
	path=${path%/}
	path=${path##*/}
	REPLY=$path
}

main() {
	"$1" "$PWD"
	local _result="$REPLY"
}

main "$@"
```



</details>



```sh
> $ for fn in slow_basename fast_basename faster_basename; do hyperfine -N --warmup 1000 --runs 1000 "./script.sh $fn"; done
Benchmark 1: ./script.sh slow_basename
  Time (mean ± σ):       3.6 ms ±   0.2 ms    [User: 2.0 ms, System: 1.4 ms]
  Range (min … max):     3.4 ms …   8.8 ms    1000 runs
 
  Warning: Statistical outliers were detected. Consider re-running this benchmark on a quiet PC without any interferences from other programs. It might help to use the '--warmup' or '--prepare' options.
 
Benchmark 1: ./script.sh fast_basename
  Time (mean ± σ):       3.2 ms ±   0.4 ms    [User: 1.6 ms, System: 1.4 ms]
  Range (min … max):     2.9 ms …  15.1 ms    1000 runs
 
  Warning: Statistical outliers were detected. Consider re-running this benchmark on a quiet PC without any interferences from other programs. It might help to use the '--warmup' or '--prepare' options.
 
Benchmark 1: ./script.sh faster_basename
  Time (mean ± σ):       2.7 ms ±   0.3 ms    [User: 1.3 ms, System: 1.2 ms]
  Range (min … max):     2.5 ms …   8.9 ms    1000 runs
 
  Warning: Statistical outliers were detected. Consider re-running this benchmark on a quiet PC without any interferences from other programs. It might help to use the '--warmup' or '--prepare' options.
```


For your questions about my self-defined "best practices"

> If you unset a var (or function), it's gone. In C you may need to then initialize the empty var to a known value, but this isn't C.

The `REPLY=` is to be absolutely certain that `REPLY` is what you think it is. For a single function, this doesn't do anything, but if you have hundreds of functions that set `REPLY`, this ensures that artifacts of running those other functions don't accidentally interfere with the current function.

The `unset -v` is necessary because we want to be absolutely sure that the variable is of the correct type - it's fundamentally more correct. See below:

<details>

<summary>script.sh</summary>

```sh
# shellcheck shell=bash

fn1() {
	unset -v REPLY
	REPLY=

	printf 'fn1: %s\n' "$(declare -p REPLY)"
	REPLY=(a b)
}

fn2() {
	# unset -v REPLY
	REPLY=

	printf 'fn2: %s\n' "$(declare -p REPLY)"
	REPLY=c
}

main() {
	fn1
	local -a result1="${REPLY[@]}"

	fn2
	local result2="$REPLY"
}

main "$@"
```



</details>

And I can't just do `unset -v REPLY` because for some reason, people like to enable `errunset`, and since I'm usually writing Bash libraries, I need to be compatible for the options that they wish to set.

> If it's function-scoped, why bother unsetting it? This is always going to print a null for the first echo, and 10 for the second:

Yeah, it's function scoped, and I still don't want that variable to leak into the callers of functions that are called later in that function like so:

```sh
otherfn() {
  local i=
  for i in {0..1}; do :; done

  foo
}
```

So by doing `unset -v i`, the goal is to have a behavior closer to scoped variables (i suppose like `let` vs `var` in JavaScript)

Reply:
> So I'm not sure what you're trying to measure when you use printf without a subshell.

In the light of day, I also have no idea what I was doing there. Good catch.

> And I can't just do unset -v REPLY because for some reason, people like to enable errunset, and since I'm usually writing Bash libraries, I need to be compatible for the options that they wish to set.

An excellent point I hadn't thought of. People love to set `-euo pipefail` because someone said it was better.

> Yeah, it's function scoped, and I still don't want that variable to leak into the callers of functions that are called later in that function like so

Again, an excellent point I hadn't thought of.

I think I'm just going to defer to your recommendations. I thought I knew shell reasonably well, but clearly I've a lot more to learn. Thank you for taking the time to demonstrate your answers with scripts.